Added new Community List page to the login flow

JakeShirley opened the pull request:
Addresses #78

<!-- 
Thank you for making a pull request! 
Since we are very busy with getting Mlem into a releaseable state, we had to introduce this short questionnaire to help us review PRs.
Before you submit your PR, please take a few minutes to fill out all the needed information.

Please note that if you do not fill out the checklist, your PR will be automatically rejected unless you are an approved contributor. 
We apologize, as we would love to dedicate the time it deserves to every PR, but at present, we are under considerable time pressure.
-->

# Checklist
- [x] I have described what this PR contains

**Choose one of the following two options:**
- - [ ] This PR does not introduce major changes
- - [x] This PR introduces major changes, and I have consulted @buresdv, @jboi or @mormaer in the *Mlem Development Matrix room*

**Choose one of the following two options:**
- - [ ] This PR does not change the UI in any way
- - [x] This PR adds new UI elements / changes the UI, and I have attached pictures or videos of the new / changed UI elements

# Pull Request Information

## About this Pull Request
The PR add an Apollo-like communities screen.  It is injected between the "Accounts" screen and the "Community View".  This is a natural "home" screen that allows people to select what they want to browse when they first get into the app.

I am brand new to Swift and SwiftUI so please bear with me!

## Screenshots and Videos
<img width="249" alt="image" src="https://github.com/buresdv/Mlem/assets/1000311/f6327651-334e-419c-acb2-6735042008ed">

Video demo: https://www.youtube.com/watch?v=R52NyvqnBug

## Additional Context
Still LOTS of clean up to do.  Still learning the ropes and seeing where I can make re-usable components and what not.

I wrong a bunch of (ugly) code to generate samples for the SwiftUI preview to function (yay!).  I rebased today (6/18) and had to disable a project setting for code coverage to get previews to work again (found the solution [here](https://stackoverflow.com/questions/59705079/swiftui-canvas-preview-compile-error-undefined-symbols-for-architecture-x86-64)).

ShadowJonathan commented on Mlem/API/Requests/Community/ListCommunities.swift:
```diff
@@ -31,6 +31,7 @@ struct ListCommunitiesRequest: APIGetRequest {
 
             .init(name: "limit", value: limit.map(String.init)),
             .init(name: "page", value: page.map(String.init)),
+            .init(name: "type_", value: type.rawValue),
```
🤦 i somehow missed these...

thanks ^^

ShadowJonathan commented on Mlem/Views/Tabs/Posts/Components/Post/Components/Community View/Community List View.swift:
```diff
@@ -0,0 +1,398 @@
+    }
+}
+
+///
+// Preview-only functions to populate the model
+// Not sure on what best practices are here
+// If these are useful these could be moved to be used externally
+///
+let fakeCommunityPrefixes: [String] =
```
Maybe add a TODO here to move this to test code, or something of the like?

ShadowJonathan reviewed (commented):
LGTM, however, when you press the "feed" button, does it now go to accounts or the list of communities to select and all?

Could we possibly have the default stack be `account -> list -> subscribed communities`?

ShadowJonathan commented:
@buresdv why was this closed?

WestonHanners commented on Mlem/Views/Tabs/Posts/Components/Post/Components/Community View/Community List View.swift:
```diff
@@ -0,0 +1,398 @@
+        ).environmentObject(favoritesTracker)
+    }
+}
+    
```
Maybe we should move other components into separate files? I won't hold the PR up for this, it's just a thought.

JakeShirley commented:
PR is ready! Moved some views out into their own files and cleaned up the UI a little bit.  Here it is:

![image](https://github.com/buresdv/Mlem/assets/1000311/9dcd6861-f7be-4458-a55d-a0f4c5676e58)
![image](https://github.com/buresdv/Mlem/assets/1000311/74a14560-6f8b-427c-89b9-2b566c07d3fe)

MrAusnadian commented:
Looks great @JakeShirley !

buresdv commented:
I can't approve this in the current state. Will give more info when out of hospital

EricBAndrews commented on Mlem/Views/Tabs/Posts/Components/Post/Components/Community List View/Community List View.swift:
```diff
@@ -0,0 +1,285 @@
+                        HomepageFeedRowView(account: account, feedType: .local, iconName: "building.2.crop.circle.fill", iconColor: Color.green, description: "Local communities from your server")
+                        HomepageFeedRowView(account: account, feedType: .all, iconName: "cloud.circle.fill", iconColor: Color.blue, description: "All communities that federate with your server")
+                        
+                        ForEach(communitySections
+                                
+                                // Only show letter headers for letters we have in our community list
+                            .filter({ (communitySection) -> Bool in
+                                getSubscriptionsAndFavorites().contains(where: { communitySection.sidebarEntry.contains(community: $0, isSubscribed: subscribedCommunities.contains($0)) })
+                            })
+                                // Only show sections which have labels to show
+                            .filter({ (communitySection) -> Bool in
+                                communitySection.inlineHeaderLabel != nil
+                            })) { communitySection in
+                                
+                                Section(header:
+                                            HStack {
+                                    Text(communitySection.inlineHeaderLabel!)
+                                    Spacer()
+                                }.id(communitySection.viewId)) {
+                                    ForEach(
+                                        getSubscriptionsAndFavorites()
+                                            .filter({ (listedCommunity) -> Bool in
+                                                // Filter down to sidebar entry which wants us
+                                                communitySection.sidebarEntry.contains(community: listedCommunity, isSubscribed: subscribedCommunities.contains(listedCommunity))
+                                            })
+                                    ) { listedCommunity in
+                                        CommuntiyFeedRowView(account: account, community: listedCommunity, subscribed: subscribedCommunities.contains(listedCommunity), communitySubscriptionChanged: self.hydrateCommunityData)
+                                    }
+                                }
+                            }
+                        
+                    }
```
Not a blocker to this merge, but ideally the data filtering happens outside of the body--that enforces nice isolation of logic and rendering, and makes both parts more legible

BrooklynMan replied:
if we can resolve conflicts on this, lets get it done and merge

JakeShirley replied:
Moved it to a utility method for now.  Definitely more cleanup to do.

EricBAndrews reviewed (approved):
Good to merge--I have left some stylistic/organizational notes, but nothing that can't wait until a follow-up PR.

BrooklynMan reviewed (approved):
well done

BrooklynMan reviewed (dismissed):
LET'S GO ALREADYYYYYYY!

BrooklynMan reviewed (approved):
well done
